app/parser/metadata_parser.py

```diff
@@ -23,77 +19,69 @@ def id_generator():
-                    attr_value = constant.parser(value)
-                    logger.debug("with value %s", attr_value)
-                elif constant.mandatory:
-                    logger.warning("Missing constant value for %s", constant.claim_id)
-                    raise ValueError("Missing constant value %s".format(constant.claim_id))
-                else:
-                    if constant.generator:
-                        logger.debug("Generating value for %s", attr_name)
-                        attr_value = constant.generator()
-                    else:
-                        logger.debug("No value provide for %s but this is not mandatory, setting to None", attr_name)
-                        attr_value = None
-                setattr(metadata, attr_name, attr_value)
-            return metadata
-        except (RuntimeError, ValueError, TypeError) as e:
-            logger.error("Unable to parse Metadata")
-            logger.exception(e)
-            raise InvalidTokenException("Incorrect data in token")
+
+    def validate(self, original_value):
+        if self.mandatory and original_value is None:
+            raise ValueError("Missing mandatory field value")
+        if self._validator:
+            return self._validator(original_value)
+
+    def generate(self):
+        if self._generator:
+            return self._generator()
+        elif not self.mandatory:
+            return None
+
+metadata_fields = {
+  "user_id": MetadataField(),
+  "ru_ref": MetadataField(),
+  "ru_name": MetadataField(),
+  "eq_id": MetadataField(),
+  "collection_exercise_sid": MetadataField(),
+  "period_id": MetadataField(),
+  "period_str": MetadataField(),
+  "ref_p_start_date": MetadataField(validator=iso_8601_data_parser),
+  "ref_p_end_date": MetadataField(validator=iso_8601_data_parser),
+  "form_type": MetadataField(),
+  "return_by": MetadataField(validator=iso_8601_data_parser),
+  "trad_as": MetadataField(mandatory=False),
+  "employment_date": MetadataField(mandatory=False, validator=iso_8601_data_parser),
+  "tx_id": MetadataField(mandatory=False, validator=uuid_4_parser, generator=id_generator),
+}
+
+
+def parse_metadata(metadata_to_check):
+    parsed = {}
+    try:
+        for key, field in metadata_fields.items():
+            logger.debug("parse_metadata: Adding attr %s", key)
+            if key in metadata_to_check:
+                attr_value = metadata_to_check[key]
+                field.validate(attr_value)
+                logger.debug("with value %s", attr_value)
+            else:
+                logger.debug("Generating value for %s", key)
+                attr_value = field.generate()
+
+            parsed[key] = attr_value
+    except (RuntimeError, ValueError, TypeError) as e:
+        logger.error("parse_metadata: Unable to parse")
+        logger.exception(e)
+        raise InvalidTokenException("Incorrect data in token")
+
+    return parsed
+
+
+def is_valid_metadata(metadata):
```

Comment:
Yeah, I wanted to rename this method (has_mandatory_metadata_fields?), but couldn't think of anything concise. It makes sense that all the validation is done in this method, rather than parse, but it'll touch lots of unit tests.